Support external and absolute URLs in import and url plugins

KidkArolis opened the pull request:
For example, it's now possible to have @import url('http://fonts.googleapis.com/css?family=Lato:300,400,700');

In addition, convert url(image.png) => require("./image.png") the way css-loader does.

The methods for testing and converting the URLs can be found here
https://github.com/webpack/loader-utils/blob/db3c7cd123355f61846497ebeef614670c075ed9/index.js#L93-L142

See the css-loader for reference
See https://github.com/webpack/css-loader/blob/70771dc2586d90439b370c6d3b5921e48058a4fe/index.js


Not sure if this is backwards compatible with the current version of rework-webpack-loader or not.

aaronj1335 commented on lib/plugins/imports.js:
```diff
@@ -1,14 +1,24 @@
+        return addImport(reworkInstance, loaderUtils.urlToRequest(url));
+      }
+    }
+    keepRules.push(rule);
```
this seems to change the logic such that the `@import` rule is left in the css even when it's translated into a `require`, which is incorrect.

aaronj1335 replied:
oh. oh nvm. i get it now.

aaronj1335 commented on lib/plugins/imports.js:
```diff
@@ -1,14 +1,24 @@
 var addImport = require('./../add-import');
 var stripQuotes = require('./../strip-quotes');
 
+var urlRe = /url\(([^)]+)\)/g;
```
this looks the same as the one in [`lib/plugins/urls.js`](https://github.com/aaronj1335/rework-webpack-loader/blob/54c5c39d3c1e92882d1d280f84c8eaaca0e92128/lib/plugins/urls.js#L13), so it should be in a shared module. i'b be cool with like a `./../patterns` module or something.

aaronj1335 commented on package.json:
```diff
@@ -8,13 +8,14 @@
   "author": "Aaron Stacy",
   "dependencies": {
-    "when": "~3.4.4",
+    "loader-utils": "^0.2.5",
```
i'm guessing you just added this via `npm install`, but please keep the lines in the previous order, so the diff only shows the line added.

aaronj1335 commented on lib/plugins/urls.js:
```diff
@@ -18,8 +19,15 @@ module.exports = function(ast, reworkInstance) {
+            addImport(reworkInstance, urlParse(url).pathname);
+            return 'url("' + mark(url) + '")';
+          } else {
+            return _;
```
i was using `_` as just a placeholder, but since it's now got some meaning, maybe rename it to `declarationValue`?

aaronj1335 commented:
hey thanks for fixing this up. i'll merge once you've got the changes in and re-publish to npm. i guess in order to be on the safe side i'll bump the major version number.

KidkArolis commented:
Thanks for the code review. Sorry for the delay - holidays ;)

I was trying to add some tests, but proved to be tricky - might come back to that later. I tried using https://github.com/webpack/extract-text-webpack-plugin to produce a standalone css file and inspect it in the test, but `extract-text-webpack-plugin` and `rework-webpack-loader` seem to be incompatible atm - the relative module imports are removed from js but not added to the standalone css file.

I've fixed up the PR based on your comments and force pushed.
I've also fixed a bug I've introduced in the previous commit where the regex with /g flag would keep internal state when `exec`ed multiple times and so only every other `import` statement would match.
This version now works well for me in my project, so _should_ be safe(ish) to merge with no tests.

aaronj1335 commented:
ugh finally merged and published to npm. i'm the worst. thanks so much for your work.

also fwiw i don't use rework with webpack anymore. the main driver was that i wanted to use webpack's css loader, and if you want to use [autoprefixer](https://github.com/postcss/autoprefixer), you have to parse everything with postcss anyway. on the flip side, you end up needing a lot less code. [here's an example](https://gist.github.com/aaronj1335/a8d7816ee28c316b3028).

should probably factor these out into modules&hellip;
